### PR TITLE
111 upload answer

### PR DESCRIPTION
**Pull request by codyseibert**
- adding new use case for filing a document (answer) and everything else that might be related to that
- refactor to use getPersistenceGateway() function instead of applicationContext.persistence.myUseCase

### CODE REVIEW

File: shared/src/business/useCases/getCases.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ exports.getCases = async ({ userId, status, applicationContext }) => {
         .getUseCases()
         .getCasesByUser({ userId, applicationContext });
     case 'irsattorney':
-      return await applicationContext.getUseCases().getCasesByIRSAttorney({
+      return await applicationContext.getUseCases().getCasesForRespondent({
```
Comment (kkoskelin) [inline shared/src/business/useCases/getCases.js:18]: We've altered the use case being called here, "getCasesForRespondent" but the `switch` statement case we're selecting is still `irsattorney`.  Is this correct?

File: efcms-service/serverless.yml [hunk 2]
```diff
@@ -289,8 +289,8 @@ resources:
             MetricsEnabled: true
 
 functions:
-  createPolicyUrl:
-    handler: src/documents/createPolicyUrl.create
+  getUploadPolicy:
+    handler: src/documents/getUploadPolicy.create
```
Comment (dknesek) [inline efcms-service/serverless.yml:293]: All S3 upload and download policy stuff is persistence related. None of it is "Document" related directly. These endpoints/lambdas are confusing. See refactoring 42.

File: efcms-service/src/applicationContext.js [hunk 3]
```diff
@@ -11,12 +11,18 @@ const { getUser } = require('ef-cms-shared/src/business/useCases/getUser');
 const { sendPetitionToIRS } = require('ef-cms-shared/src/business/useCases/sendPetitionToIRS');
 const { updateCase } = require('ef-cms-shared/src/business/useCases/updateCase');
 const { uploadCasePdfs } = require('ef-cms-shared/src/business/useCases/uploadCasePdfs');
-const { getCasesByIRSAttorney } = require('ef-cms-shared/src/business/useCases/getCasesByIRSAttorney');
+const { getCasesForRespondent } = require('ef-cms-shared/src/business/useCases/respondent/getCasesForRespondent');
+const { getUploadPolicy } = require('ef-cms-shared/src/persistence/getUploadPolicy');
```
Comment (dknesek) [inline efcms-service/src/applicationContext.js]: This should go away. It's not a use case but a regular function that is part of persistence.

File: web-client/src/main/actions/viewDocument.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+export default async ({ applicationContext, props }) => {
+  const {
+    url,
+  } = await applicationContext.getPersistenceGateway().getDownloadPolicyUrl({
+    documentId: props.documentId,
+    applicationContext,
+  });
+  window.open(url, '_blank', 'noopener noreferrer');
```
Comment (kkoskelin) [inline web-client/src/main/actions/viewDocument.js]: the window features should be comma-separated, e.g. `'noopener,noreferrer`'

File: efcms-service/src/applicationContext.js [hunk 5]
```diff
@@ -11,12 +11,18 @@ const { getUser } = require('ef-cms-shared/src/business/useCases/getUser');
...
+const { getDownloadPolicyUrl } = require('ef-cms-shared/src/persistence/getDownloadPolicyUrl');
```
Comment (dknesek) [inline efcms-service/src/applicationContext.js]: This should go away. It's not a use case but a regular function that is part of persistence.

File: efcms-service/src/applicationContext.js [hunk 6]
```diff
@@ -11,12 +11,18 @@ const { getUser } = require('ef-cms-shared/src/business/useCases/getUser');
...
 
 module.exports = {
-  persistence: {
-    ...awsDynamoPersistence,
-    ...awsS3Persistence
+  getPersistenceGateway: () => {
+    return {
+      ...awsDynamoPersistence,
+      ...awsS3Persistence,
+      getUploadPolicy,
```
Comment (dknesek) [inline efcms-service/src/applicationContext.js:44]: These should be private function in the gateway. No one else should call them.

File: efcms-service/src/applicationContext.js [hunk 7]
```diff
@@ -11,12 +11,18 @@ const { getUser } = require('ef-cms-shared/src/business/useCases/getUser');
...
+      getDownloadPolicyUrl,
```
Comment (dknesek) [inline efcms-service/src/applicationContext.js:45]: Ditto

File: web-client/src/main/sequences/submitDocument.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import unsetFormSubmitting from '../actions/unsetFormSubmitting';
+import formatSearchParams from '../actions/formatSearchParams';
+import uploadDocument from '../actions/uploadDocument';
+
+export default [
+  setFormSubmitting,
+  clearAlerts,
+  uploadDocument,
+  unsetFormSubmitting,
+  // getCreateCaseAlertSuccess,
```
Comment (kkoskelin) [inline web-client/src/main/sequences/submitDocument.js]: is this a `TODO` item, or should it be removed?

File: web-client/src/tests/caseJourney.test.js [hunk 9]
```diff
@@ -33,14 +33,13 @@ describe('Tax payer', async () => {
   });
   fakeFile.name = 'requestForPlaceOfTrial.pdf';
 
-  describe('Initiate case', () => {
+  describe.only('Initiate case', () => {
```
Comment (kkoskelin) [inline web-client/src/tests/caseJourney.test.js]: oops, let's remove the `only`?

**Review by kkoskelin — CHANGES_REQUESTED** [review]
That's a monster! Happy to discuss my requested changes/questions.

File: shared/src/business/useCases/respondent/fileAnswer.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+}) => {
+  //validate the pdf
+  if (!answerDocument) {
+    throw new UnprocessableEntityError(
+      'answer document cannot be null or invalid',
+    );
+  }
+
+  //get upload policy
+  const policy = await applicationContext
```
Comment (ericsorenson) [inline shared/src/business/useCases/respondent/fileAnswer.js]: The use case shouldn't be aware of upload policies, this should be in the persistence object.

File: shared/src/business/useCases/getCases.js [hunk 11]
```diff
@@ -15,7 +15,7 @@ exports.getCases = async ({ userId, status, applicationContext }) => {
         .getUseCases()
         .getCasesByUser({ userId, applicationContext });
     case 'irsattorney':
```
Comment (dknesek) [inline shared/src/business/useCases/getCases.js]: We need to split this use case interactor. See refactoring 50.

File: shared/src/business/useCases/respondent/fileAnswer.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    documentId: answerDocumentId,
+    userId: userId,
+  };
+  const caseWithAnswer = new Case({
+    ...caseToUpdate,
+    documents: [...caseToUpdate.documents, answerDocumentMetadata],
+  });
+  caseWithAnswer.validateWithError(new UnprocessableEntityError());
+
+  return await applicationContext.getUseCases().updateCase({
```
Comment (dknesek) [inline shared/src/business/useCases/respondent/fileAnswer.js:37]: See refactoring 52.

File: web-client/src/main/sequences/submitDocument.js (hunk 8 above)
Comment (ericsorenson) [inline web-client/src/main/sequences/submitDocument.js]: can this comment be deleted?

File: web-client/src/router.js [hunk 13]
```diff
@@ -29,6 +29,10 @@ const router = {
       document.title = `File a petition ${pageTitleSuffix}`;
       app.getSequence('gotoFilePetition')();
     });
+    route('/file-a-document', () => {
```
Comment (ericsorenson) [inline web-client/src/router.js:36]: Filing a document doesn't have its own page.

File: web-client/src/tests/caseJourney.test.js (hunk 9 above)
Comment (ericsorenson) [inline web-client/src/tests/caseJourney.test.js]: Please remove the `only`

File: web-client/src/applicationContexts/dev.js [hunk 14]
```diff
@@ -17,7 +28,12 @@ const applicationContext = {
     return process.env.API_URL || 'http://localhost:3000/v1';
   },
   getPersistenceGateway: () => {
-    return persistenceGateway;
+    return {
+      uploadPdf,
+      getDownloadPolicyUrl,
```
Comment (dknesek) [inline web-client/src/applicationContexts/dev.js]: delete

File: web-client/src/applicationContexts/dev.js [hunk 15]
```diff
@@ -17,7 +28,12 @@ const applicationContext = {
...
+      getUploadPolicy,
```
Comment (dknesek) [inline web-client/src/applicationContexts/dev.js]: delete

File: web-client/src/applicationContexts/prod.js [hunk 16]
```diff
@@ -17,7 +28,12 @@ const applicationContext = {
     return process.env.API_URL || 'http://localhost:8080';
   },
   getPersistenceGateway: () => {
-    return awsPersistenceGateway;
+    return {
+      uploadPdf,
+      getDownloadPolicyUrl,
```
Comment (dknesek) [inline web-client/src/applicationContexts/prod.js]: delete

File: web-client/src/applicationContexts/prod.js [hunk 17]
```diff
@@ -17,7 +28,12 @@ const applicationContext = {
...
+      getUploadPolicy,
```
Comment (dknesek) [inline web-client/src/applicationContexts/prod.js]: delete

**Review by dknesek — CHANGES_REQUESTED** [review]

File: efcms-service/src/applicationContext.js (hunk 5 above)
Comment (codyseibert) [reply efcms-service/src/applicationContext.js]: We do not have these as use cases.  They are currently persistence functions which we return in the getPersistenceGateway() factory.

File: efcms-service/src/applicationContext.js (hunk 3 above)
Comment (codyseibert) [reply efcms-service/src/applicationContext.js]: We do not have these as use cases.  They are currently persistence functions which we return in the getPersistenceGateway() factory.

**Review by kkoskelin — APPROVED** [review]

File: shared/src/business/useCases/respondent/fileAnswer.js (hunk 12 above)
Comment (codyseibert) [reply shared/src/business/useCases/respondent/fileAnswer.js:37]: will address in future

**Review by dknesek — APPROVED** [review]

**Review by ericsorenson — APPROVED** [review]

**Pull request merged**